test/e2e/federated-service.go

```diff
@@ -0,0 +1,209 @@
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+const (
+	UserAgentName = "federation-e2e-service-controller"
+	// TODO(madhusudancs): Using the same values as defined in the federated
+	// service controller. Replace it with the values from the e2e framework.
+	KubeAPIQPS   = 20.0
+	KubeAPIBurst = 30
+
+	FederatedServiceTimeout = 5 * time.Minute
+
+	FederatedServiceName = "federated-service"
+	FederatedServicePod  = "federated-service-test-pod"
+
+	// TODO: Only suppoprts IPv4 addresses. Also add a regexp for IPv6 addresses.
+	FederatedIPAddrRegexp  = `(([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])`
+	FederatedDNS1123Regexp = `([a-z0-9]([-a-z0-9]*[a-z0-9])?\.)*([a-z0-9]([-a-z0-9]*[a-z0-9])?)`
+)
+
+var _ = framework.KubeDescribe("Service [Feature:Federation]", func() {
+	var clusterClientSets []*release_1_3.Clientset
+	f := framework.NewDefaultFederatedFramework("service")
+
+	BeforeEach(func() {
+		contexts := f.GetUnderlyingFederatedContexts()
+
+		for _, context := range contexts {
+			framework.Logf("Creating cluster object: %s (%s)", context.Name, context.Cluster.Cluster.Server)
+			cluster := federation.Cluster{
+				ObjectMeta: api.ObjectMeta{
+					Name: context.Name,
+				},
+				Spec: federation.ClusterSpec{
+					ServerAddressByClientCIDRs: []federation.ServerAddressByClientCIDR{
+						{
+							ClientCIDR:    "0.0.0.0/0",
+							ServerAddress: context.Cluster.Cluster.Server,
+						},
+					},
+				},
+			}
+			_, err := f.FederationClientset.Federation().Clusters().Create(&cluster)
+			framework.ExpectNoError(err, "Creating cluster")
+		}
+
+		By("Obtaining a list of all the clusters")
+		clusterList, err := f.FederationClientset.Federation().Clusters().List(api.ListOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		framework.Logf("%d clusters found", len(clusterList.Items))
+
+		for _, cluster := range clusterList.Items {
+			framework.Logf("Creating a clientset for the cluster %s", cluster.Name)
+
+			Expect(framework.TestContext.KubeConfig).ToNot(Equal(""), "KubeConfig must be specified to load clusters' client config")
+			kubecfg, err := clientcmd.LoadFromFile(framework.TestContext.KubeConfig)
+			framework.ExpectNoError(err, "error loading KubeConfig: %v", err)
+
+			cfgOverride := &clientcmd.ConfigOverrides{
+				ClusterInfo: clientcmdapi.Cluster{
+					Server: cluster.Spec.ServerAddressByClientCIDRs[0].ServerAddress,
+				},
+			}
+			ccfg := clientcmd.NewNonInteractiveClientConfig(*kubecfg, cluster.Name, cfgOverride, clientcmd.NewDefaultClientConfigLoadingRules())
+			cfg, err := ccfg.ClientConfig()
+			Expect(err).NotTo(HaveOccurred())
+
+			cfg.QPS = KubeAPIQPS
+			cfg.Burst = KubeAPIBurst
+			clset := release_1_3.NewForConfigOrDie(restclient.AddUserAgent(cfg, UserAgentName))
+			clusterClientSets = append(clusterClientSets, clset)
+		}
+	})
+
+	It("should be able to discover a federated service", func() {
+		framework.SkipUnlessFederated(f.Client)
```

Comment:
You should also skip all of the BeforeEach() stuff above (cluster registration etc) unless we're federated.